gateway/api/v1/serializers.py

```diff
@@ -43,6 +45,30 @@ def validate(self, attrs):
                 "At least one of attributes (entrypoint, image) is required."
             )
 
+        # validate dependencies
+        # allowlist stored in json config file (eventually via configmap)
+        # sample:
+        # allowlist = { "wheel": ["0.44.0", "0.43.2"] }
+        # where the values for each key are allowed versions of dependency
+        deps = json.loads(attrs.get("dependencies", None))
+        with open(settings.GATEWAY_ALLOWLIST_CONFIG, encoding="utf-8", mode="r") as f:
+            allowlist = json.load(f)
+
+        # If no allowlist specified, all dependencies allowed
+        if len(allowlist.keys()) > 0:
+            for d in deps:
+                dep, ver = d.split("==")
```

Comment:
How do you see here to force always to pin a version if you want to add a dependency?

I mean, check if the user sends: `wheel>=x.y.z`. So check if `dependencies` includes `>=` and if so raise a `ValidationError` to pin it. In that case we have one less corner to take care.

Reply:
I was not aware a user was allowed to specify anything other than `==` when passing dependencies

Reply:
well, it seems like the whole suite of [specifications](https://pip.pypa.io/en/stable/reference/requirement-specifiers/) is allowed....

Reply:
mmm... I didn't think about this until I saw your PR to be honest, sorry. Correct me if I'm wrong but basically if they use `SomeProject >= 1.2, < 2.0` we will return:

```
"Dependency SomeProject >= 1.2, < 2.0 is not allowed"
```

Maybe we can touch a little bit the message to say: `Dependency {dep} must be pinned or is not allowed`. Or something like that? Because I see a bit difficult to try to catch all the scenarios.

Reply:
yeah, we should probably force the user to either (a) specify a specific version, or (b) don't specify a version at all. the more esoteric cases (like version >3.0.1, < 4) are probably more work than it's worth to implement

Reply:
Actually, we may be able to pull in the Ray logic to handle dependency parsing, need to do a little more investigation.

@Tansito since this PR is just getting the framework in place and thus not implementing any user-facing changes by populating the allowlist, let's merge this as is and then handle fine-tuning the dependency versioning in a follow up PR (that'll help keep the changes in each PR manageable)

Reply:
Sure, let me just give a last fast review taking that into account then.